torch/_dynamo/compile_package.py

```diff
@@ -0,0 +1,756 @@
+            "static_input_indices",
+            "enable_log",
+            "pre_dispatch",
+        ]
+        pickle.dump(
+            {field: getattr(graph_compile.aot_config, field) for field in fields}, f
+        )
+
+    with open(
+        os.path.join(save_path, f"{graph_compile.name}.compile_fx_kwargs"), "wb"
+    ) as f:
+        cfk = graph_compile.compile_fx_kwargs.copy()
+        assert "package" in cfk
+        cfk["package"] = None
+        pickle.dump(cfk, f)
+
+    assert graph_compile.forward_aoti is not None
+    src = graph_compile.forward_aoti.filename
+    dst = os.path.join(save_path, f"{graph_compile.name}.forward.pt2")
+    if isinstance(src, str) and src.endswith(".pt2"):
+        # Already packed aoti file.
+        shutil.copy(src, dst)
+    else:
+        # Unpacked aoti files.
+        assert isinstance(src, list)
+        torch._inductor.package.package_aoti(dst, {"forward": src})
+
+    assert graph_compile.backward_aoti is None, "TODO NYI"
+
+
+class SerializedGuard:
+    source: Source
+    create_fn: Callable[["GuardBuilderBase", "Guard"], None]
+    metadata: object
+
+    def __init__(self, guard):
+        self.source = guard.originating_source
+        if isinstance(guard.create_fn, functools.partial):
+            self.create_fn = guard.create_fn.func
+        else:
+            self.create_fn = guard.create_fn
+        self.metadata = None
+
+
+def _construct_raw_module(state):
+    mod = torch.nn.Module()
+    mod.__setstate__(state)
+    return mod
+
+
+class ScopePickler(pickle.Pickler):
+    def reducer_override(self, obj):
+        if isinstance(obj, torch.Tensor):
+            return torch.tensor, ([],)  # TODO use FakeTensor?
+        elif isinstance(obj, torch.nn.Module):
+            if obj.__class__.__getstate__ == torch.nn.Module.__getstate__:
+                return _construct_raw_module, (obj.__getstate__(),)
+
+        if type(obj).__qualname__ != type(obj).__name__:
+            raise RuntimeError(
+                f"Type {type(obj)} for object {obj} cannot be saved "
+                + "into torch.compile() package since it's defined in local scope. "
+                + "Please define the class at global scope (top level of a module)."
+            )
+
+        return NotImplemented
+
+
+def serialize_local_scope(scope) -> bytes:
+    buf = io.BytesIO()
+    ScopePickler(buf).dump(scope)
+    return buf.getvalue()
+
+
+@dataclass
+class GuardsState:
+    guards: list[SerializedGuard]
+    global_state: str
+    f_code: SerializedCode  # TODO deduplicate this.
+    input_source_to_sizes_strides: dict[Source, tuple[torch.Size, tuple[int, ...]]]
+    local_scope: bytes
+    torch_function_mode_stack: bytes
+
+
+# TODO Potentially GuardBuilder can be split into the first and
+#      second half, and serialier should read data from the return
+#      value of the first half.
+#      For prototyping we just leave this as a separate class.
+class GuardSerializer:
```

Comment:
What do we think about moving the guard serde into its own PR? That way you could test it separately, and have it be generic, rather than specific to the package API